work_queue/src/work_queue_worker.c

```diff
@@ -78,6 +78,12 @@ typedef enum {
 	CONTAINER_MODE_UMBRELLA
 } container_mode_t;
 
+struct environment {
```

Comment:
Let's get more verbose about names.
struct `wq_conda_environment`: if it can really only be used for conda.
struct `wq_software_environment`: if it has potential use outside of conda.

Reply:
Re: renaming. Yes, that's much better.

Re: the separation of specify_file. The protocol actions are separate. The tarball is still sent with infile. With regards to the API, this is another wrinkle. First I tried it separate, but it did not feel natural, as you would have to specify the file twice (once to send it, once to mark it as a the environment). Also, you would have to specify a remote name for the tarball, which would never be used in the task space.

To solve the above, I tried adding a WORK_QUEUE_TARBALL type, that let you specify an input file as a taball, with a remote name as a directory to expand to. This did not work so well, because not all tar files are expanded equally. In conda environments there is no directory containing all subdirectories, and you have to: mkdir mydir; tar -C mydir tarball.tgz. For almost any other tarball, you would expect: mkdir mydir; tar -C mydir --strip-components=1 tarball.tgz, as they usually have a directory that contains all. 

With that, the tarball infile message would need to contain the argument to strip-components, or we would have to create a message specially for tarballs.

As it is right now, there is the advantage that a user has no expectation on where the environment would be mounted, only that it will be activated. This is important, as it is unpacked for all tasks, and not for a particular remote name directory that a user may specify.

In summary, it was more complicated than it first appeared!

Reply:
Ok, that makes sense to me.  Probably safer to have a "narrow" API that works in one way, rather than multiple touchpoints that can be misconfigured...